### PR TITLE
Add `Diff` and `Spread` types

### PR DESCRIPTION
**Pull request by EdwardDrapkin**

### CODE REVIEW

**Comment by sindresorhus** [comment]
I just added some contribution guidelines. Would you be able to review them? I'm also interested in feedback if anything there is unclear or could be improved.

https://github.com/sindresorhus/type-fest/blob/master/.github/contributing.md

File: index.d.ts [hunk 1]
```diff
@@ -84,3 +84,23 @@ const ab: Merge<Foo, Bar> = {a: 1, b: 2};
 ```
 */
 export type Merge<FirstType, SecondType> = Omit<FirstType, Extract<keyof FirstType, keyof SecondType>> & SecondType;
+		  
+/**
+ * Diffs two objects.
```
Comment (sindresorhus) [inline index.d.ts]: Can you add this to the readme too?

File: index.d.ts [hunk 2]
```diff
@@ -84,3 +84,23 @@ const ab: Merge<Foo, Bar> = {a: 1, b: 2};
...
+ *
+ * Given objects with types T and V, returns an object that has all the keys in T that do not also exist in V.
+ *
+ * @example `type Safe = Diff<AllProperties, UnsafeProperties>`
+ */
+export type Diff<T extends {}, V extends {}> = {
```
Comment (sindresorhus) [inline index.d.ts]: Can you use more descriptive type names than `T` and `V`? See the other types for inspiration.

File: index.d.ts [hunk 3]
```diff
@@ -84,3 +84,23 @@ const ab: Merge<Foo, Bar> = {a: 1, b: 2};
 ```
 */
 export type Merge<FirstType, SecondType> = Omit<FirstType, Extract<keyof FirstType, keyof SecondType>> & SecondType;
+		  
+/**
+ * Diffs two objects.
+ *
+ * Given objects with types T and V, returns an object that has all the keys in T that do not also exist in V.
+ *
+ * @example `type Safe = Diff<AllProperties, UnsafeProperties>`
```
Comment (sindresorhus) [inline index.d.ts]: Can you include a more comprehensive example? Should also include the import statement.

File: index.d.ts [hunk 4]
```diff
@@ -84,3 +84,23 @@ const ab: Merge<Foo, Bar> = {a: 1, b: 2};
...
+ */
+export type Diff<T extends {}, V extends {}> = {
+    [P in Exclude<keyof T, keyof V>]: T[P];
+};
+
+		  
+/**
+ * Returns a type modeling the result of spreading two objects together.
+ *
+ * Given objects with types T and V, returns an object that has a type that represents {...T, ...V}
```
Comment (sindresorhus) [inline index.d.ts]: How is this different from `Merge`?

File: index.d.ts [hunk 5]
```diff
@@ -84,3 +84,23 @@ const ab: Merge<Foo, Bar> = {a: 1, b: 2};
...
+ * @example `const a: Spread<X, Y> = { ...x, ...y }`
```
Comment (sindresorhus) [inline index.d.ts]: Doesn't TS already infer the new type correctly from object-spread? Maybe I'm missing the use-case for this type.

**Comment by sindresorhus** [comment]
Ping :)

**Comment by sindresorhus** [comment]
@CvX @BendingBender Do you think these types are something we should add?

**Comment by CvX** [comment]
I prefer to extract types from existing code. I find it difficult to judge if given type will find a use in the future. 😅 Some real-world examples would go a long way.

**Comment by BendingBender** [comment]
@sindresorhus I'm not sure about this PR. Like @CvX I have difficulties imagining using types like this. They are clever solutons but usually a simpler "dumb" solution is easier to understand and is at the same time obvious for every dev who reads it. I'm sure that these types can be useful, especially in the context of ambient declarations but I have yet to encounter a use case for them.

As you've already pointed out, this library is mostly for types that are useful for your projects (https://github.com/sindresorhus/type-fest/issues/3#issuecomment-472474036). So what's the point adding types that you can't even imagine to use?

**Comment by alixaxel** [comment]
I think I have a use case for where this would be helpful. In a ORM I built, I have a few decorators that can potentially be attached for a model, and will automatically populate `id` + `createdAt` + `updatedAt`.

However, in some models, it doesn't make sense to have either one of those:

```typescript
Omit<T, keyof (Identified & Timestamped)>
&
PartialMaybe<T, Identified>
&
PartialMaybe<T, Timestamped>)
```

For context:

```typescript
type PartialMaybe<T, O> = T extends O ? Partial<O> : {};

interface Identified {
  id: string;
}

interface Timestamped {
  createdAt: firestore.Timestamp;
  updatedAt?: firestore.Timestamp;
}
```

So in practice this will conditionally (if they exist) make `Identified` and `Timestamped` types optional.

**Comment by Richienb** [comment]
@sindresorhus My use case for `Spread` is when I'm trying to document return types without creating new interfaces.

```ts
interface returnTypes {
	somePropertyToExtend: string
}

function someFunction(): Spread<returnTypes, {
	somethingElseThatIsNeeded: string
}>

function anotherFunction(): Spread<returnTypes, {
	anotherThingThatIsNeeded: string
}>
```

**Comment by sindresorhus** [comment]
It would be better to open a new issue to discuss this.

**Pull request closed without merging**